### DISCUSSION TITLE
Rust std fs slower than Python!? No, it's hardware!

### DISCUSSION THREAD

**Discussion by giscus** [Xuanwo/blog#416] [Announcements]
# Rust std fs slower than Python!? No, it's hardware!

An infrastructure engineer, focused on distributed storage system

https://xuanwo.io/2023/04-rust-std-fs-slower-than-python/

**Comment by wooparadog** [comment]
CPU bug must be so difficult to debug. Good job locating the root cause and it's a great read.

**Comment by lilydjwg** [comment]
A little story about how I found the offset difference.

"Two identical syscalls can't have different timings!" I thought while staring at the two `strace -T` log files, after ruling out the difference of `length` argument.

Some process-wide setting? I wanted to search the memory address, and took some time to figure out the `-e raw=read` argument. And suddenly, I noticed the memory addresses passed to `read` had different offsets!

**Reply by Xuanwo** [reply]
Truly fantastic!

**Comment by michaelmior** [comment]
Great writeup! One minor point is that the crate that should be used for `jemallactor` is actually `tikv-jemallocator`. (Both work, but the first is deprecated.)

**Reply by Xuanwo** [reply]
Thanks for the suggestion! I've been discussing this with upstream before, but there hasn't been any progress yet: https://github.com/tikv/jemallocator/issues/63

**Comment by bgolinvaux** [comment]
Very interesting!   Kudos for this thorough analysis.  You and your friends deserve a lot of respect!

**Comment by rod-stuchi** [comment]
What amazing discovery you made, it must had took a lot of time and effort to accomplish it.
It was a pleasure to read, it's objective, right to the point. 👏

**Comment by tbo007** [comment]
Thanks for the interesting article. One question: how did the two strace outputs (rust / phython) gave you the idea to delve deeper into mmap? (...I suspect that mmap...) Thanks.

**Reply by Xuanwo** [reply]
@lilydjwg shared her story here: https://github.com/Xuanwo/blog/discussions/416#discussioncomment-7704455

**Comment by bokrenok** [comment]
How about `musl libc` instead of `gibc`?
Would there also be such difference with usage of `rep movsb` asm instructions?

**Reply by Xuanwo** [reply]
The tests haven't been conducted yet. The outcome is closely linked to the way `musl` and `glibc` use the `rep mov` instructions. If `musl` isn't reliant on these, I believe we might see a different result.

**Reply by jan-hudec** [reply]
The copy is from kernel, so it's done by `copy_to_user`, not `memcpy` and libc shouldn't make any difference.

**Comment by ykoehler** [comment]
What I do not get, is that if Python is faster than Rust/C on the same hardware, would not that indicate that Python found a work-around to the issue and C/Rust is not using it?  If Python runs faster without the ucode change, then C/Rust should use the same approach on AMD CPU as to get equal or faster results when the CPU ucode is bad, no?

**Comment by Blindinlights** [comment]
https://tech.ahrefs.com/skylake-bug-a-detective-story-ab1ad2beddcd
A similar story,Intel CPU bug detected by OCaml runtime crash detection

**Comment by jan-hudec** [comment]
It would be useful to test whether the problem is caused by alignment of the destination or the alignment of source and destination being similar. Because if the former, a workaround could be implemented in the memory copy functions by copying the first few bytes separately. While for the later, adding random padding to *large* allocations could help.

**Comment by EvanCarroll** [comment]
Why not just patch LLVM such that on that architecture when built with glibc it provides an offset to fread?

**Reply by jan-hudec** [reply]
LLVM isn't the right place to do it, LLVM just emits calls to library functions. The place to add a workaround is the kernel and the standard library—though I think Linux kernel already exposes CPU-optimized memcpy for the standard library to use, so might be just kernel. But we need to know a bit more about which case is the pathological one before a generic workaround could be proposed.

Also note that it only is fread in this specific case, but the problem really affects any memcpy, copy_to_user and copy_from_user with parameters of specific properties (the later two are kernel functions that transfer data between kernel and userspace for system calls).

**Reply by EvanCarroll** [reply]
@jan-hudec great response, thanks.

**Comment by Xuanwo** [comment]
Fixed in glibc: https://sourceware.org/git/gitweb.cgi?p=glibc.git;h=0c0d39fe4aeb0f69b26e76337c5dfd5530d5d44e

**Reply by EvanCarroll** [reply]
Look forward to the update, re the blog entry,

> Our only hope is to address this issue in glibc by disabling FSRM as necessary. Progress has been made on the glibc front: [x86: Improve ERMS usage on Zen3](https://inbox.sourceware.org/libc-alpha/20231031200925.3297456-3-adhemerval.zanella@linaro.org/T/#m4f4458b8b16b0ee2bac1578f0383b10eee2312d8). Stay tuned for updates.
